phabricatoremails/settings.py

```diff
@@ -0,0 +1,122 @@
+import configparser
+import pathlib
+import smtplib
+from configparser import ConfigParser
+from dataclasses import dataclass
+from logging import Logger
+from typing import Any
+
+from phabricatoremails.db import Db
+from phabricatoremails.logging import create_dev_logger, create_logger
+from phabricatoremails.mail import SesMail, SmtpMail, FsMail
+from phabricatoremails.progress import StaticProgress, PhabricatorProgress
+from phabricatoremails.source import FileSource, PhabricatorSource
+from sqlalchemy import create_engine
+
+
+def _parse_logger(config: ConfigParser):
+    """Provides a logger set up according to our configuration
+
+    The regular logger is noisy and implements the MozLog JSON format, so we also
+    have a "dev" logger that simply prints out each log without any annotations.
+    The correct logger is chosen based on the provided configuration file.
+    """
+    if config.has_section("dev"):
+        return create_dev_logger()
+    else:
+        return create_logger()
+
+
+def _parse_source(config: ConfigParser, logger: Logger):
+    """Provides data-fetching implementations according to our configuration
+
+    To facilitate easier local development, we have a few different combinations of
+    ways to get data (Phabricator, or just a local file) as well as how we want
+    to perform at runtime (fetch-and-send emails once, or on a continuous loop).
+    Returns a "data source" and "progress" implementation
+    """
+    dev_source_file = config.get("dev", "file", fallback=None)
+    if dev_source_file:
+        return FileSource(pathlib.Path(dev_source_file).resolve()), StaticProgress(None)
+
+    host = config.get("phabricator", "host")
+    token = config.get("phabricator", "token")
+    poll_gap_seconds = int(config.get("phabricator", "poll_gap_seconds", fallback="60"))
+    override_since_key = config.get("dev", "since_key", fallback=None)
+    override_limit = config.get("dev", "limit", fallback=None)
+    if host.endswith("/"):
+        host = host[:-1]
+
+    source = PhabricatorSource(host, token, override_limit)
+    progress = (
+        StaticProgress(override_since_key)
+        if override_since_key
+        else PhabricatorProgress(logger, poll_gap_seconds)
+    )
```

Comment:
A lot is happening here, can you refactor this to be a little more clear, and add some comments to help understand the if statement.